### ISSUE TITLE
Build error with ASan and UBSan on macOS

### ISSUE THREAD

**Issue by nth10sd** [facebook/hermes#270]
Build error:

```
[501/505] Linking CXX executable bin/fuzzer-jsi-entry
FAILED: bin/fuzzer-jsi-entry
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -fno-sanitize-recover=undefined -fvisibility=hidden -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -fno-sanitize-recover=undefined -fsanitize=fuzzer tools/fuzzers/CMakeFiles/fuzzer-jsi-entry.dir/fuzzer-jsi-entry.cpp.o  -o bin/fuzzer-jsi-entry  API/hermes/libhermesapi.a  jsi/libjsi.a  lib/VM/libhermesVMRuntime.a  lib/ADT/libhermesADT.a  lib/Platform/libhermesPlatform.a  lib/BCGen/HBC/libhermesHBCBackend.a  lib/Inst/libhermesInst.a  lib/BCGen/libhermesBackend.a  lib/libhermesFrontend.a  lib/FrontEndDefs/libhermesFrontEndDefs.a  lib/libhermesOptimizer.a  lib/SourceMap/libhermesSourceMap.a  lib/Parser/libhermesParser.a  lib/AST/libhermesAST.a  lib/Support/libhermesSupport.a  lib/Regex/libhermesRegex.a  lib/Platform/Unicode/libhermesPlatformUnicode.a  external/dtoa/libdtoa.a  -framework  CoreFoundation  external/llvh/lib/Support/libLLVHSupport.a  external/llvh/lib/Demangle/libLLVHDemangle.a  -licucore && :
ld: file not found: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.3/lib/darwin/libclang_rt.fuzzer_osx.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[502/505] Linking CXX shared library API/hermes/libhermes.dylib
ld: warning: direct access in function 'hermes::vm::NativeFunction::_nativeCall(hermes::vm::NativeFunction*, hermes::vm::Runtime*)' from file 'lib/VM/libhermesVMRuntime.a(Callable.cpp.o)' to global weak symbol 'typeinfo for hermes::vm::CallResult<hermes::vm::HermesValue, (hermes::vm::detail::CallResultSpecialize)2> (void*, hermes::vm::Runtime*, hermes::vm::NativeArgs)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'hermes::vm::NativeFunction::_nativeCall(hermes::vm::NativeFunction*, hermes::vm::Runtime*)' from file 'lib/VM/libhermesVMRuntime.a(Callable.cpp.o)' to global weak symbol 'typeinfo for hermes::vm::CallResult<hermes::vm::HermesValue, (hermes::vm::detail::CallResultSpecialize)2> (void*, hermes::vm::Runtime*, hermes::vm::NativeArgs)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'void hermes::vm::GCBase::clearEntriesWithUnreachableKeys<bool (*)(hermes::vm::GCCell const*)>(hermes::vm::GenGC*, hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, bool (*)(hermes::vm::GCCell const*))' from file 'lib/VM/libhermesVMRuntime.a(GenGCNC.cpp.o)' to global weak symbol 'typeinfo for bool (hermes::vm::GCCell const*)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'void hermes::vm::GCBase::clearEntriesWithUnreachableKeys<bool (*)(hermes::vm::GCCell const*)>(hermes::vm::GenGC*, hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, bool (*)(hermes::vm::GCCell const*))' from file 'lib/VM/libhermesVMRuntime.a(GenGCNC.cpp.o)' to global weak symbol 'typeinfo for bool (hermes::vm::GCCell const*)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'bool hermes::vm::GCBase::markFromReachableWeakMapKeys<hermes::vm::CompleteMarkState::FullMSCMarkTransitiveAcceptor, bool (*)(hermes::vm::GCCell const*), hermes::vm::GenGC::completeWeakMapMarking()::$_0>(hermes::vm::GenGC*, hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, hermes::vm::CompleteMarkState::FullMSCMarkTransitiveAcceptor&, llvm::DenseMap<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, std::__1::list<hermes::vm::detail::WeakRefKey*, std::__1::allocator<hermes::vm::detail::WeakRefKey*> >, llvm::DenseMapInfo<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*>, llvm::detail::DenseMapPair<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, std::__1::list<hermes::vm::detail::WeakRefKey*, std::__1::allocator<hermes::vm::detail::WeakRefKey*> > > >*, bool (*)(hermes::vm::GCCell const*), hermes::vm::GenGC::completeWeakMapMarking()::$_0)::'lambda'(hermes::vm::detail::WeakRefKey*)::operator()(hermes::vm::detail::WeakRefKey*) const' from file 'lib/VM/libhermesVMRuntime.a(GenGCNC.cpp.o)' to global weak symbol 'typeinfo for bool (hermes::vm::GCCell const*)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'bool hermes::vm::GCBase::markFromReachableWeakMapKeys<hermes::vm::CompleteMarkState::FullMSCMarkTransitiveAcceptor, bool (*)(hermes::vm::GCCell const*), hermes::vm::GenGC::completeWeakMapMarking()::$_0>(hermes::vm::GenGC*, hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, hermes::vm::CompleteMarkState::FullMSCMarkTransitiveAcceptor&, llvm::DenseMap<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, std::__1::list<hermes::vm::detail::WeakRefKey*, std::__1::allocator<hermes::vm::detail::WeakRefKey*> >, llvm::DenseMapInfo<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*>, llvm::detail::DenseMapPair<hermes::vm::JSWeakMapImpl<(hermes::vm::CellKind)43>*, std::__1::list<hermes::vm::detail::WeakRefKey*, std::__1::allocator<hermes::vm::detail::WeakRefKey*> > > >*, bool (*)(hermes::vm::GCCell const*), hermes::vm::GenGC::completeWeakMapMarking()::$_0)::'lambda'(hermes::vm::detail::WeakRefKey*)::operator()(hermes::vm::detail::WeakRefKey*) const' from file 'lib/VM/libhermesVMRuntime.a(GenGCNC.cpp.o)' to global weak symbol 'typeinfo for bool (hermes::vm::GCCell const*)' from file 'API/hermes/CMakeFiles/libhermes.dir/hermes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
[505/505] Linking CXX executable bin/hdb
ninja: build stopped: subcommand failed.
```

I had run `utils/build/configure.py --enable-asan --enable-ubsan`, the corresponding Linux counterpart was fine.

Running on macOS 10.15.5

**Comment by nth10sd** [comment]
Missing the `libclang_rt.fuzzer_osx.a` file:

```
$ ls /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.3/lib/darwin/
libclang_rt.asan_ios_dynamic.dylib		libclang_rt.osx.a				libclang_rt.ubsan_iossim.a
libclang_rt.asan_iossim_dynamic.dylib		libclang_rt.profile_ios.a			libclang_rt.ubsan_iossim_dynamic.dylib
libclang_rt.asan_osx_dynamic.dylib		libclang_rt.profile_iossim.a			libclang_rt.ubsan_osx.a
libclang_rt.asan_tvos_dynamic.dylib		libclang_rt.profile_osx.a			libclang_rt.ubsan_osx_dynamic.dylib
libclang_rt.asan_tvossim_dynamic.dylib		libclang_rt.profile_tvos.a			libclang_rt.ubsan_tvos.a
libclang_rt.asan_watchos_dynamic.dylib		libclang_rt.profile_tvossim.a			libclang_rt.ubsan_tvos_dynamic.dylib
libclang_rt.asan_watchossim_dynamic.dylib	libclang_rt.profile_watchos.a			libclang_rt.ubsan_tvossim.a
libclang_rt.cc_kext.a				libclang_rt.profile_watchossim.a		libclang_rt.ubsan_tvossim_dynamic.dylib
libclang_rt.cc_kext_driverkit.a			libclang_rt.tsan_iossim_dynamic.dylib		libclang_rt.ubsan_watchos.a
libclang_rt.cc_kext_ios.a			libclang_rt.tsan_osx_dynamic.dylib		libclang_rt.ubsan_watchos_dynamic.dylib
libclang_rt.cc_kext_tvos.a			libclang_rt.tsan_tvossim_dynamic.dylib		libclang_rt.ubsan_watchossim.a
libclang_rt.cc_kext_watchos.a			libclang_rt.tvos.a				libclang_rt.ubsan_watchossim_dynamic.dylib
libclang_rt.driverkit.a				libclang_rt.ubsan_ios.a				libclang_rt.watchos.a
libclang_rt.ios.a				libclang_rt.ubsan_ios_dynamic.dylib
```

**Comment by mhorowitz** [comment]
It appears that Xcode is just missing the necessary libraries to use the fuzzer sanitizer.

We'll conditionalize this as a real fix, but if you need a short-term workaround, you can just comment out the contents of `hermes/tools/fuzzers/CMakeLists.txt` and run the build again.

**Comment by neildhar** [comment]
Thank you for reporting this issue, it should be fixed in 8c481e8c8c6534fad42a74e145cd16e19f941710
